### PR TITLE
[Refactoring] Redux and log pattern related refactoring

### PR DESCRIPTION
**Pull request by mengweieric**
### Description
- Use createSelector and connect together
- Log pattern modularization
- Fix all related liniting errors

Just combined redux and log pattern refactoring together in one PR. 

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/framework/redux/store/index.ts [hunk 1]
```diff
@@ -38,12 +38,12 @@ const store = configureStore({
 
 if (process.env.NODE_ENV === 'development' && module.hot) {
   module.hot.accept('./rootReducer', () => {
-    store.replaceReducer(persistReducer(persistConfig, require('./rootReducer').default));
+    store.replaceReducer(persistReducer(persistConfig, require('./rootReducer').default)); // eslint-disable-line @typescript-eslint/no-var-requires
```
Comment (mengweieric) [inline public/framework/redux/store/index.ts:41]: Will change this line to conform liniting rules, just wanted to keep this PR with fewer unrelated changes.

File: public/components/event_analytics/redux/slices/count_distribution_slice.ts [hunk 2]
```diff
@@ -26,6 +26,9 @@ export const countDistributionSlice = createSlice({
 
 export const { render } = countDistributionSlice.actions;
 
-export const selectCountDistribution = (state) => state.countDistribution;
+export const selectCountDistribution = createSelector(
+  (state) => state.countDistribution,
+  (countDisState) => countDisState
```
Comment (derek-ho) [inline public/components/event_analytics/redux/slices/count_distribution_slice.ts:31]: what does this second line do, that just takes an argument and returns it

File: public/components/event_analytics/redux/slices/count_distribution_slice.ts [hunk 3]
```diff
@@ -26,6 +26,9 @@ export const countDistributionSlice = createSlice({
...
+);
```
[the remark this answers is not in the document]
Comment (mengweieric) [reply public/components/event_analytics/redux/slices/count_distribution_slice.ts:32]: mostly yes, but with pattern we can create different types of creators that return different state value as needed in a clean way as well as leverage its memorization.

File: public/components/event_analytics/redux/slices/count_distribution_slice.ts (hunk 2 above)
Comment (mengweieric) [reply public/components/event_analytics/redux/slices/count_distribution_slice.ts:31]: Yes as for this selectCountDistribution, I only need the entire countDisState

**Review by derek-ho — APPROVED** [review]

File: public/components/event_analytics/explorer/explorer.tsx [hunk 4]
```diff
@@ -1466,6 +1311,7 @@ export const Explorer = ({
         explorerData,
         http,
         query,
+        notifications,
```
Comment (derek-ho) [inline public/components/event_analytics/explorer/explorer.tsx:1314]: ?

File: public/components/event_analytics/explorer/log_patterns/patterns_header.tsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                  Cancel
+                </EuiButtonEmpty>
+              </EuiFlexItem>
+              <EuiFlexItem grow={false}>
+                <EuiButton
+                  size="s"
+                  fill
+                  onClick={async () => {
+                    setIsPatternConfigPopoverOpen(false);
+                    onPatternApply();
```
Comment (derek-ho) [inline public/components/event_analytics/explorer/log_patterns/patterns_header.tsx:100]: await?

File: public/components/event_analytics/explorer/explorer.tsx (hunk 4 above)
Comment (mengweieric) [reply public/components/event_analytics/explorer/explorer.tsx:1314]: The notification is used by log pattern after modularization, and for easy access to all components that needs to access notification,  I added to context

File: public/components/event_analytics/explorer/log_patterns/patterns_header.tsx (hunk 5 above)
Comment (mengweieric) [reply public/components/event_analytics/explorer/log_patterns/patterns_header.tsx:100]: Thanks for the catch, I will remove async here as onPatternApply is an async func in upper scope.

File: public/components/event_analytics/explorer/explorer.tsx (hunk 4 above)
Comment (joshuali925) [reply public/components/event_analytics/explorer/explorer.tsx:1314]: Curious are there plans to expose CoreStart/CoreSetup services as context or js services instead of passing objects to functions? 

In code, objects like `http`, `pplService` are passed down everywhere. it's tedious and becomes messy, and sometimes it's not easy or not possible to pass the dependency in arguments (for example inside an embeddable component)

Comment (mengweieric) [reply public/components/event_analytics/explorer/explorer.tsx:1314]: Yes, it's in the plan, one to-do item on my end is to clear up services that are supposed to be exposed as context.

**Review by ps48 — APPROVED** [review]
